### ISSUE TITLE
Default learning rate warmup steps

### ISSUE THREAD

**Issue by francois-rozet** [ClashLuke/HeavyBall#29]
All optimizers have `warmup_steps=0` by default except `SOAP` and `PSGD` that have `warmup_steps=1`. I don't understand why. Looking into `chainable.py`, it seems like it makes no difference, as at the first step (`step = 0`), `step` is replaced with `step + 1` before the warmup check, which results in `step < group['warmup_steps']` to always be false.

https://github.com/ClashLuke/HeavyBall/blob/b579f13a2329ee239ab5ec48132949436d9a4116/heavyball/chainable.py#L466-L469

I would suggest re-writing the code as

```python
        if group['warmup_steps'] and step < group['warmup_steps']:
            group['prev_lr'] = group['lr'] = group['base_lr'] * (step + 1) / (group['warmup_steps'] + 1)
        else:
            group['prev_lr'] = group['lr'] = group['base_lr']

        group['step'] = state['step'] = step + 1
```

and setting `warmup_steps` to 0 for `SOAP` and `PSGD`. I would actually suggest removing `warmup_steps` altogether in the future, to ensure compatibility with PyTorch LR schedulers.

Note the use of $\frac{n + 1}{N + 1}$ instead of $\frac{n}{N}$ to prevent singularity at $n = 0$.

**Comment by ClashLuke** [comment]
Interesting, thank you for digging into it! I wasn't aware of that discrepancy.\
You're right; phasing out `warmup_steps` is probably the best option. The `foreach=True` optimizers can already deal with changing learning rates, but it's very experimental. - I'll leave this issue open until `warmup_steps` is removed.

Regarding the off-by-one error: `step` is initialized to 0 and incremented by one before its first use, as all internal functions expect a step starting at 1.

**Comment by francois-rozet** [comment]
Ok, then I would suggest the following code.

```python
        group['step'] = state['step'] = step = step + 1
        group['prev_lr'] = group['lr'] = group['base_lr'] * step / max(step, group['warmup_steps'] + 1)
```

If we remove the last `+ 1`, `warmup_steps=1` does not lead to any warmup step.
I can make a PR applying these changes as well as setting `warmup_steps=0` for all optimizers if you want.

**Comment by ClashLuke** [comment]
Elegant! That'd also pair well with improving the torch.Tensor step compatibility (necessary to compile the entire step)
